packages/eventual-send/src/handled-promise.js

```diff
@@ -246,11 +247,6 @@ export const makeHandledPromise = () => {
         // Remove stale pending handlers, set to canonical form.
         shorten(handledP);
 
-        // Ensure our pendingHandler is cleaned up if not already.
-        if (promiseToPendingHandler.has(handledP)) {
-          handledP.then(_ => promiseToPendingHandler.delete(handledP));
```

Comment:
`promiseToPendingHandler` is a WeakMap so it shouldn't be too leaky.

But if we attach just a `.then` handler and handledP is rejected, this expression produces an unhandled rejection.  If 
 we handle it with a `.catch`, then we silence legitimate warnings if `handledP` isn't ever handled by userspace.

Reply:
An alternative would be:
```js
const cleanup = _ => { promiseToPendingHandler.delete(handledP); }
Promise.resolve().then(() => handledP).then(cleanup , cleanup);
```

That way you don't silence the original handledP, and you cleanup in both settlement paths.

Reply:
> An alternative would be:

Really cool!  Thanks for that.

Reply:
Oh, but it doesn't work for me.  The following doesn't trigger the `unhandledRejection` callback when it should:

```js
const h = Promise.reject(Error('expected to be uncaught'));
Promise.resolve().then(() => h).then(() => {}, () => {});
```

Reply:
Oh right because the initial promise was actually handled in the `then` assimilation. Only option in that case is to re-create a promise which won't be handled:

```js
const h = Promise.reject(Error('expected to be uncaught'));
// Create a new promise which won't be handled
Promise.resolve().then(() => h);
h.then(() => {}, () => {});
```

Reply:
I'm not satisfied with any of these answers.  Either they:
a) unconditionally suppress the unhandled rejection even if `handledP` doesn't have a `.catch` handler attached by the user; or
b) create an additional unhandled rejection that cannot be caught (because it falls out of scope), even if the original `handledP` has a `.catch` handler attached by the user.

I have tried but failed to find a solution that avoids both of these problems.  Am I missing something obvious?

The problem to solve is to come up with YOUR-CODE-HERE manipulating `h` such that:
```js
// Should result in "unhandled rejection" event
h = Promise.reject('unhandled'); ...YOUR-CODE-HERE
// Should NOT result in "unhandled rejection" event
h = Promise.reject('handled'); h.catch(() => {}); ...YOUR-CODE-HERE
```

It is my assertion that the only `YOUR-CODE-HERE` that can work is something that doesn't register any handlers on `h`--neither `.then`, `.catch`, or `.finally`.

Reply:
Yeah it looks like the reason `.finally` worked when I tried it is because it creates a new promise that was itself unhandled.

The requirement to not chain is indeed incompatible with unhandled rejection tracking. If we can't put a harden before resolving the promise to the user (as in before in the chain vs before in time relying on promise reactions ordering), then yes there is no way.

Reply:
I think there is a more general question too whether a generic handled promise should work in a non-lockdown environment, in which case the explicit hardening should only be done as a lockdown of HandledPromise.